## 1. Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

## 2. Flat file digestion
STRUCTURED DATASET FROM CDC WONDER

In [2]:
# load csv file
df = pd.read_csv('../data/raw/YEAR_GENDER_AGE_1999_2020.csv')

## 3. Inspect and validate

In [3]:
# check data
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1147 entries, 0 to 1146
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Notes                      135 non-null    str    
 1   Year                       1078 non-null   float64
 2   Year Code                  1078 non-null   float64
 3   Sex                        1056 non-null   str    
 4   Sex Code                   1056 non-null   str    
 5   Five-Year Age Groups       1012 non-null   str    
 6   Five-Year Age Groups Code  1012 non-null   str    
 7   Deaths                     1079 non-null   str    
 8   Population                 1079 non-null   str    
 9   Crude Rate                 1079 non-null   str    
dtypes: float64(2), str(8)
memory usage: 139.4 KB


In [4]:
# check first 5 rows
df.head()

,Notes,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
0,NaN,1999.0,1999.0,Female,F,< 1 year,1,Suppressed,1852123,Suppressed
1,NaN,1999.0,1999.0,Female,F,1-4 years,1-4,Suppressed,7493614,Suppressed
2,NaN,1999.0,1999.0,Female,F,5-9 years,5-9,Suppressed,10052618,Suppressed
3,NaN,1999.0,1999.0,Female,F,10-14 years,10-14,Suppressed,9855664,Suppressed
4,NaN,1999.0,1999.0,Female,F,15-19 years,15-19,44,9761569,0.5


In [5]:
# check last 5 rows
df.tail()

,Notes,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
1142,were available at the time of release.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1143,10. The population figures used in the calcula...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1144,resident population that is under one year of ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1145,11. Changes to cause of death classification a...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1146,http://wonder.cdc.gov/wonder/help/mcd.html#ICD...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Check missing values

In [6]:
df.isna().sum()

Notes                        1012
Year                           69
Year Code                      69
Sex                            91
Sex Code                       91
Five-Year Age Groups          135
Five-Year Age Groups Code     135
Deaths                         68
Population                     68
Crude Rate                     68
dtype: int64

### Check duplicate rows

In [7]:
df.duplicated().sum()

np.int64(4)

In [8]:
df[df.duplicated(keep = False)]

,Notes,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
1079,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1091,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1093,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1095,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1100,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Check count of unique values across all columns

In [9]:
df.nunique()

Notes                         65
Year                          22
Year Code                     22
Sex                            2
Sex Code                       2
Five-Year Age Groups          23
Five-Year Age Groups Code     23
Deaths                       575
Population                   860
Crude Rate                   233
dtype: int64

### Check unique values in `Notes`

In [10]:
# return no. of unique values in 'Notes' column
df['Notes'].nunique()

65

In [11]:
# return list of unique values in 'Notes' column
list(df['Notes'].unique())

[nan,
 'Total',
 '---',
 'Dataset: Multiple Cause of Death, 1999-2020',
 'Query Parameters:',
 'Title: YEAR_GENDER_AGE_1999_2020',
 'MCD - ICD-10 Codes: T40.0 (Opium); T40.1 (Heroin); T40.2 (Other opioids); T40.3 (Methadone); T40.4 (Other synthetic narcotics);',
 'T40.6 (Other and unspecified narcotics)',
 'Group By: Year; Sex; Five-Year Age Groups',
 'Show Totals: True',
 'Show Zero Values: True',
 'Show Suppressed: True',
 'Calculate Rates Per: 100,000',
 'Rate Options: Default intercensal populations for years 2001-2009 (except Infant Age Groups)',
 'Help: See http://wonder.cdc.gov/wonder/help/mcd.html for more information.',
 'Query Date: Jun 2, 2026 10:24:01 PM',
 'Suggested Citation: Centers for Disease Control and Prevention, National Center for Health Statistics. National Vital Statistics',
 'System, Mortality 1999-2020 on CDC WONDER Online Database, released in 2021. Data are from the Multiple Cause of Death Files,',
 '1999-2020, as compiled from data provided by the 57 vital 

In [12]:
# return data where 'Notes' column has text
df[df['Notes'].notna()]

,Notes,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
23,Total,1999.0,1999.0,Female,F,NaN,NaN,2189,142237295,1.5
47,Total,1999.0,1999.0,Male,M,NaN,NaN,6458,136802873,4.7
48,Total,1999.0,1999.0,NaN,NaN,NaN,NaN,8647,279040168,3.1
72,Total,2000.0,2000.0,Female,F,NaN,NaN,2464,143368343,1.7
96,Total,2000.0,2000.0,Male,M,NaN,NaN,6582,138053563,4.8
...,...,...,...,...,...,...,...,...,...,...
1142,were available at the time of release.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1143,10. The population figures used in the calcula...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1144,resident population that is under one year of ...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1145,11. Changes to cause of death classification a...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [13]:
# return indexes of rows where 'Notes' column has text
df[df['Notes'].notna()].index

Index([  23,   47,   48,   72,   96,   97,  121,  145,  146,  170,
       ...
       1137, 1138, 1139, 1140, 1141, 1142, 1143, 1144, 1145, 1146],
      dtype='int64', length=135)

In [14]:
# return indexes of rows where 'Notes' column is '---`
df[(df['Notes'] == '---')].index

Index([1079, 1091, 1093, 1095, 1100], dtype='int64')

In [15]:
# return index of first row where 'Notes' column has text
df[df['Notes'] == '---'].index[0]

np.int64(1079)

### Drop metadata

In [16]:
# assign first index where 'Notes' has text to `cutoff`
cutoff = df[df['Notes'] == '---'].index[0]
cutoff

np.int64(1079)

In [17]:
# return df_working excluding rows where 'Notes' column has text
# check last 5 rows
df_working = df.iloc[:cutoff]
df_working.tail()

,Notes,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
1074,NaN,2020.0,2020.0,Male,M,100+ years,100+,0,Not Applicable,Not Applicable
1075,NaN,2020.0,2020.0,Male,M,Not Stated,NS,Suppressed,Not Applicable,Not Applicable
1076,Total,2020.0,2020.0,Male,M,NaN,NaN,49718,162256202,30.6
1077,Total,2020.0,2020.0,NaN,NaN,NaN,NaN,70222,329484123,21.3
1078,Total,NaN,NaN,NaN,NaN,NaN,NaN,585507,6746356647,8.7


**ADD TO REUSABLE FUNCTION**

In [18]:
# check df_working
df_working.info()

<class 'pandas.DataFrame'>
RangeIndex: 1079 entries, 0 to 1078
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Notes                      67 non-null     str    
 1   Year                       1078 non-null   float64
 2   Year Code                  1078 non-null   float64
 3   Sex                        1056 non-null   str    
 4   Sex Code                   1056 non-null   str    
 5   Five-Year Age Groups       1012 non-null   str    
 6   Five-Year Age Groups Code  1012 non-null   str    
 7   Deaths                     1079 non-null   str    
 8   Population                 1079 non-null   str    
 9   Crude Rate                 1079 non-null   str    
dtypes: float64(2), str(8)
memory usage: 128.1 KB


### Checkin missing values

In [19]:
df_working.isna().sum()

Notes                        1012
Year                            1
Year Code                       1
Sex                            23
Sex Code                       23
Five-Year Age Groups           67
Five-Year Age Groups Code      67
Deaths                          0
Population                      0
Crude Rate                      0
dtype: int64

In [20]:
# return number of rows whre 'Notes' is 'Total'
len(df_working[df_working['Notes'] == 'Total'])

67

In [21]:
# drop rows where 'Notes' is 'Total'
df_working = df_working[df_working['Notes'] != 'Total']

**ADD TO REUSABLE FUNCTION**

In [22]:
# check df_working
df_working.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Notes                      0 non-null      str    
 1   Year                       1012 non-null   float64
 2   Year Code                  1012 non-null   float64
 3   Sex                        1012 non-null   str    
 4   Sex Code                   1012 non-null   str    
 5   Five-Year Age Groups       1012 non-null   str    
 6   Five-Year Age Groups Code  1012 non-null   str    
 7   Deaths                     1012 non-null   str    
 8   Population                 1012 non-null   str    
 9   Crude Rate                 1012 non-null   str    
dtypes: float64(2), str(8)
memory usage: 128.9 KB


In [23]:
# check missing values again
df_working.isna().sum()

Notes                        1012
Year                            0
Year Code                       0
Sex                             0
Sex Code                        0
Five-Year Age Groups            0
Five-Year Age Groups Code       0
Deaths                          0
Population                      0
Crude Rate                      0
dtype: int64

In [24]:
# check duplicates again
df_working.duplicated().sum()

np.int64(0)

### Drop unnecessary columns

In [25]:
df_working = df_working.drop(columns = 'Notes')

**ADD TO REUSABLE FUNCTION**

In [26]:
# check df_working
df_working.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       1012 non-null   float64
 1   Year Code                  1012 non-null   float64
 2   Sex                        1012 non-null   str    
 3   Sex Code                   1012 non-null   str    
 4   Five-Year Age Groups       1012 non-null   str    
 5   Five-Year Age Groups Code  1012 non-null   str    
 6   Deaths                     1012 non-null   str    
 7   Population                 1012 non-null   str    
 8   Crude Rate                 1012 non-null   str    
dtypes: float64(2), str(7)
memory usage: 120.8 KB


## 4. Create reusable function to remove metadata

In [27]:
def remove_metadata(filepath: str) -> pd.DataFrame:
    
    # load file
    df = pd.read_csv(filepath, dtype=str)
    
    # assign index of first row with metadata to cutoff
    cutoff = df[df['Notes'] == '---'].index[0]
    
    # drop rows with metadata
    df_sans_metadata = df.iloc[:cutoff]

    # drop rows where 'Notes' == 'Total'
    df_sans_metadata = df_sans_metadata[df_sans_metadata['Notes'] != 'Total']
    
    # drop 'Notes' column
    df_sans_metadata = df_sans_metadata.drop(columns='Notes')
    
    return df_sans_metadata

# load single file
filepath = '../data/raw/YEAR_GENDER_AGE_1999_2020.csv'
df_clean = remove_metadata(filepath)

print(f"Shape: {df_clean.shape}")
print(f"Year range: {df_clean['Year'].min()} – {df_clean['Year'].max()}")

# write to processed directory
df_clean.to_csv(
    '../data/processed/YEAR_GENDER_AGE_1999_2020_preprocessed.csv', 
    index = False
)

print('Saved: ../data/processed/YEAR_GENDER_AGE_1999_2020_preprocessed.csv')

Shape: (1012, 9)
Year range: 1999 – 2020
Saved: ../data/processed/YEAR_GENDER_AGE_1999_2020_preprocessed.csv


### Confirm file

In [28]:
os.path.exists('../data/processed/YEAR_GENDER_AGE_1999_2020_preprocessed.csv')

True

## 5. Inspect and validate

In [29]:
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 9 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Year                       1012 non-null   str  
 1   Year Code                  1012 non-null   str  
 2   Sex                        1012 non-null   str  
 3   Sex Code                   1012 non-null   str  
 4   Five-Year Age Groups       1012 non-null   str  
 5   Five-Year Age Groups Code  1012 non-null   str  
 6   Deaths                     1012 non-null   str  
 7   Population                 1012 non-null   str  
 8   Crude Rate                 1012 non-null   str  
dtypes: str(9)
memory usage: 129.0 KB


In [30]:
df_clean.head()

,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
0,1999,1999,Female,F,< 1 year,1,Suppressed,1852123,Suppressed
1,1999,1999,Female,F,1-4 years,1-4,Suppressed,7493614,Suppressed
2,1999,1999,Female,F,5-9 years,5-9,Suppressed,10052618,Suppressed
3,1999,1999,Female,F,10-14 years,10-14,Suppressed,9855664,Suppressed
4,1999,1999,Female,F,15-19 years,15-19,44,9761569,0.5


In [31]:
df_clean.tail()

,Year,Year Code,Sex,Sex Code,Five-Year Age Groups,Five-Year Age Groups Code,Deaths,Population,Crude Rate
1071,2020,2020,Male,M,85-89 years,85-89,18,Not Applicable,Not Applicable
1072,2020,2020,Male,M,90-94 years,90-94,16,Not Applicable,Not Applicable
1073,2020,2020,Male,M,95-99 years,95-99,Suppressed,Not Applicable,Not Applicable
1074,2020,2020,Male,M,100+ years,100+,0,Not Applicable,Not Applicable
1075,2020,2020,Male,M,Not Stated,NS,Suppressed,Not Applicable,Not Applicable


In [32]:
df_clean.isna().sum()

Year                         0
Year Code                    0
Sex                          0
Sex Code                     0
Five-Year Age Groups         0
Five-Year Age Groups Code    0
Deaths                       0
Population                   0
Crude Rate                   0
dtype: int64

In [33]:
df_clean.duplicated().sum()

np.int64(0)

In [34]:
df_clean.nunique()

Year                          22
Year Code                     22
Sex                            2
Sex Code                       2
Five-Year Age Groups          23
Five-Year Age Groups Code     23
Deaths                       508
Population                   793
Crude Rate                   223
dtype: int64

## 6. Column validation and cleaning

### `Year` and `YEAR CODE`

In [35]:
# check unique values in 'Year'
df_clean['Year'].unique()

<ArrowStringArray>
['1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
 '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
 '2017', '2018', '2019', '2020']
Length: 22, dtype: str

In [36]:
# check unique values in 'Year Code'
df_clean['Year Code'].unique()

<ArrowStringArray>
['1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
 '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
 '2017', '2018', '2019', '2020']
Length: 22, dtype: str

In [37]:
# check if 'Year' == 'Year Code'
(df_clean['Year'] == df_clean['Year Code']).all()

np.True_

In [38]:
# convert 'Year' column to numerical values
df_clean['Year'] = df_clean['Year'].astype('int64')

**ADD TO REUSABLE FUNCTION**

In [39]:
# drop 'Year Code' column
df_clean.drop(columns = ['Year Code'], inplace = True)

**ADD TO REUSABLE FUNCTION**

In [40]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 8 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Year                       1012 non-null   int64
 1   Sex                        1012 non-null   str  
 2   Sex Code                   1012 non-null   str  
 3   Five-Year Age Groups       1012 non-null   str  
 4   Five-Year Age Groups Code  1012 non-null   str  
 5   Deaths                     1012 non-null   str  
 6   Population                 1012 non-null   str  
 7   Crude Rate                 1012 non-null   str  
dtypes: int64(1), str(7)
memory usage: 112.9 KB


### `Sex`

In [41]:
# check unique values in 'Sex'
df_clean['Sex'].unique()

<ArrowStringArray>
['Female', 'Male']
Length: 2, dtype: str

In [42]:
# check unique values in 'Sex Code'
df_clean['Sex Code'].unique()

<ArrowStringArray>
['F', 'M']
Length: 2, dtype: str

In [43]:
# drop 'Sex Code' column
df_clean.drop(columns = ['Sex Code'], inplace = True)

**ADD TO REUSABLE FUNCTION**

In [44]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Year                       1012 non-null   int64
 1   Sex                        1012 non-null   str  
 2   Five-Year Age Groups       1012 non-null   str  
 3   Five-Year Age Groups Code  1012 non-null   str  
 4   Deaths                     1012 non-null   str  
 5   Population                 1012 non-null   str  
 6   Crude Rate                 1012 non-null   str  
dtypes: int64(1), str(6)
memory usage: 103.9 KB


### `Five-Year Age Groups` and `Five-Year Age Groups Code`

In [45]:
# check unique values in 'Five-Year Age Groups'
df_clean['Five-Year Age Groups'].unique()

<ArrowStringArray>
[    '< 1 year',    '1-4 years',    '5-9 years',  '10-14 years',
  '15-19 years',  '20-24 years',  '25-29 years',  '30-34 years',
  '35-39 years',  '40-44 years',  '45-49 years',  '50-54 years',
  '55-59 years', '60-64 years ',  '65-69 years',  '70-74 years',
  '75-79 years',  '80-84 years',  '85-89 years',  '90-94 years',
  '95-99 years',   '100+ years',   'Not Stated']
Length: 23, dtype: str

In [46]:
# check unique values in 'Five-Year Age Groups Code'
df_clean['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [47]:
# temporarily strip 'Five-Year Age Groups' to match 'Five-Year Age Groups Code'
(df_clean['Five-Year Age Groups']
    .str.replace('< ', '')
    .str.replace(' years?', '', regex = True)
    .str.replace('Not Stated', 'NS')
    .str.strip()
    .unique()
)

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [48]:
# check both columns against each other
((df_clean['Five-Year Age Groups']
    .str.replace('< ', '')
    .str.replace(' years?', '', regex = True)
    .str.replace('Not Stated', 'NS')
    .str.strip()
) == df_clean['Five-Year Age Groups Code']).all()

np.True_

In [49]:
# drop 'Five-Year Age Groups' column
df_clean.drop(columns = ['Five-Year Age Groups'], inplace = True)

**ADD TO REUSABLE FUNCTION**

In [50]:
# check unique values in 'Five-Year Age Groups' column
df_clean['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [51]:
# check unique values in 'Five-Year Age Groups' column
df_clean['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [52]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 6 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Year                       1012 non-null   int64
 1   Sex                        1012 non-null   str  
 2   Five-Year Age Groups Code  1012 non-null   str  
 3   Deaths                     1012 non-null   str  
 4   Population                 1012 non-null   str  
 5   Crude Rate                 1012 non-null   str  
dtypes: int64(1), str(5)
memory usage: 85.4 KB


### `Deaths`

In [53]:
# check total number of unique values in 'Deaths' column
df_clean['Deaths'].nunique()

508

In [54]:
# check first five rows in 'Deaths' column
df_clean['Deaths'].head()

0    Suppressed
1    Suppressed
2    Suppressed
3    Suppressed
4            44
Name: Deaths, dtype: str

In [55]:
# check last five rows in 'Deaths' column
df_clean['Deaths'].tail()

1071            18
1072            16
1073    Suppressed
1074             0
1075    Suppressed
Name: Deaths, dtype: str

In [56]:
# convert values in 'Deaths' column into numerical values
df_clean['Deaths'] = df_clean['Deaths'].astype('int64')

ValueError: invalid literal for int() with base 10: 'Suppressed'

**UNEXPECTED ERROR - SOME VALUES CANNOT BE CONVERTED INTO NUMERICAL VALUE**

In [57]:
# convert values in 'Deaths' column to numerical values, otherwise converts to NaN
pd.to_numeric(df_clean['Deaths'], errors = 'coerce')

0        NaN
1        NaN
2        NaN
3        NaN
4       44.0
        ... 
1071    18.0
1072    16.0
1073     NaN
1074     0.0
1075     NaN
Name: Deaths, Length: 1012, dtype: float64

In [58]:
# returns values from 'Deaths' column where values = NaN (False)
pd.to_numeric(df_clean['Deaths'], errors = 'coerce').isna()

0        True
1        True
2        True
3        True
4       False
        ...  
1071    False
1072    False
1073     True
1074    False
1075     True
Name: Deaths, Length: 1012, dtype: bool

In [59]:
# return rows where 'Deaths' = NaN
df_clean[pd.to_numeric(df_clean['Deaths'], errors = 'coerce').isna()].head()

,Year,Sex,Five-Year Age Groups Code,Deaths,Population,Crude Rate
0,1999,Female,1,Suppressed,1852123,Suppressed
1,1999,Female,1-4,Suppressed,7493614,Suppressed
2,1999,Female,5-9,Suppressed,10052618,Suppressed
3,1999,Female,10-14,Suppressed,9855664,Suppressed
19,1999,Female,90-94,Suppressed,Not Applicable,Not Applicable


**THERE ARE OBJECTS IN THE `Deaths` COLUMN WHICH PREVENT NUMERICAL CONVERSION**

In [60]:
# return unique values in 'Deaths' column where NaN is False
(df_clean[pd.to_numeric(df_clean['Deaths'], errors = 'coerce')
    .isna()]['Deaths']
    .unique()
)

<ArrowStringArray>
['Suppressed']
Length: 1, dtype: str

**`Suppressed` PREVENTS NUMERICAL CONVERSION**

In [61]:
# return sum of 'Suppressed' occurrences in 'Deaths' column
(df_clean['Deaths'] == 'Suppressed').sum()

np.int64(231)

**TOTAL OF 231 INSTANCES WHERE 'Deaths' is 'Suppressed'**

In [62]:
# add 'Deaths_Supressed' column before replacing 'Suppressed' with NaN
df_clean['Deaths_Suppressed'] = df_clean['Deaths'] == 'Suppressed'

**ADD TO REUSABLE FUNCTION**

In [63]:
# check unique values in `Deaths_Suppressed` column
df_clean['Deaths_Suppressed'].unique()

array([ True, False])

In [64]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Year                       1012 non-null   int64
 1   Sex                        1012 non-null   str  
 2   Five-Year Age Groups Code  1012 non-null   str  
 3   Deaths                     1012 non-null   str  
 4   Population                 1012 non-null   str  
 5   Crude Rate                 1012 non-null   str  
 6   Deaths_Suppressed          1012 non-null   bool 
dtypes: bool(1), int64(1), str(5)
memory usage: 86.3 KB


In [65]:
# return values in 'Deaths' colummn where 'Deaths' is 'Suppressed'
df_clean.loc[df_clean['Deaths'] == 'Suppressed', 'Deaths']

0       Suppressed
1       Suppressed
2       Suppressed
3       Suppressed
19      Suppressed
           ...    
1031    Suppressed
1049    Suppressed
1055    Suppressed
1073    Suppressed
1075    Suppressed
Name: Deaths, Length: 231, dtype: str

In [66]:
# assign NaN to 'Suppressed' values in 'Deaths' column 
df_clean.loc[df_clean['Deaths'] == 'Suppressed', 'Deaths'] = np.nan

**ADD TO REUSABLE FUNCTION**

In [67]:
# return sum of NaN occurences in 'Deaths' column
df_clean['Deaths'].isna().sum()

np.int64(231)

In [68]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype
---  ------                     --------------  -----
 0   Year                       1012 non-null   int64
 1   Sex                        1012 non-null   str  
 2   Five-Year Age Groups Code  1012 non-null   str  
 3   Deaths                     781 non-null    str  
 4   Population                 1012 non-null   str  
 5   Crude Rate                 1012 non-null   str  
 6   Deaths_Suppressed          1012 non-null   bool 
dtypes: bool(1), int64(1), str(5)
memory usage: 84.1 KB


In [69]:
# convert values in 'Deaths' column into numerical values
df_clean['Deaths'] = pd.to_numeric(df_clean['Deaths'], errors='coerce')
df_clean['Deaths'].dtype

dtype('float64')

**ADD TO REUSABLE FUNCTION**

In [70]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 1012 entries, 0 to 1075
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       1012 non-null   int64  
 1   Sex                        1012 non-null   str    
 2   Five-Year Age Groups Code  1012 non-null   str    
 3   Deaths                     781 non-null    float64
 4   Population                 1012 non-null   str    
 5   Crude Rate                 1012 non-null   str    
 6   Deaths_Suppressed          1012 non-null   bool   
dtypes: bool(1), float64(1), int64(1), str(4)
memory usage: 81.8 KB


### `Population`

In [71]:
# convert values in 'Population' column into numerical values
df_clean['Population'] = df_clean['Population'].astype('int64')

ValueError: invalid literal for int() with base 10: 'Not Applicable'

**UNEXPECTED ERROR - SOME VALUES CANNOT BE CONVERTED INTO NUMERICAL VALUE**

In [72]:
# return unique values in 'Population' column where NaN is False
(df_clean[pd.to_numeric(df_clean['Population'], errors = 'coerce')
    .isna()]['Population']
    .unique()
)

<ArrowStringArray>
['Not Applicable']
Length: 1, dtype: str

**`Not Applicable` PREVENTS NUMERICAL CONVERSION**

In [73]:
# return sum of 'Not Applicable' occurrences in 'Population' column
(df_clean['Population'] == 'Not Applicable').sum()

np.int64(220)

**TOTAL OF 220 INSTANCES WHERE 'Population' is 'Not Applicable'**

In [74]:
# check if 'Deaths_Suppressed' overlaps with 'Not Applicable' in the 'Population' column
df_clean[df_clean['Population'] == 'Not Applicable']['Deaths_Suppressed'].value_counts()

Deaths_Suppressed
True     117
False    103
Name: count, dtype: int64

**Not Applicable' DOES NOT OVERLAP WITH 'Deaths_Suppressed'**

In [75]:
# check if unique values of 'Five-Year Age Groups Code' overlaps with 
# 'Not Applicable' in the 'Population' column
df_clean[df_clean['Population'] == 'Not Applicable']['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
['85-89', '90-94', '95-99', '100+', 'NS']
Length: 5, dtype: str

**'Not Applicable' DOES OVERLAP WITH THE FOLLOWING 'Five-Year Age Groups'**

In [76]:
# count total rows from the 'Population' column
# where 'Five-Year Age Groups Code' meets the following age groups
(df_clean[df_clean['Five-Year Age Groups Code']
    .isin(['85-89', '90-94', '95-99', '100+', 'NS'])]['Population']
    .value_counts()
)

Population
Not Applicable    220
Name: count, dtype: int64

**TOTAL OF 220 INSTANCES WHERE 'Five-Year Age Groups Code' OVERLAP WITH 'Not Applicable' from 'Population'**

In [77]:
# check count of 'Deaths' where 'Population' is 'Not Applicable'
df_clean[df_clean['Population'] == 'Not Applicable']['Deaths'].value_counts(dropna=False)

Deaths
NaN     117
0.0      50
11.0      6
10.0      5
21.0      5
16.0      5
22.0      5
12.0      4
23.0      3
15.0      3
28.0      3
19.0      3
20.0      2
30.0      2
24.0      1
14.0      1
27.0      1
13.0      1
32.0      1
17.0      1
18.0      1
Name: count, dtype: int64

In [78]:
# return row(s) where 'Population' is 'Not Applicable' and 'Deaths' > 0.0
df_clean[
    (df_clean['Population'] == 'Not Applicable') & 
    (df_clean['Deaths'] > 0.0)
]

,Year,Sex,Five-Year Age Groups Code,Deaths,Population,Crude Rate,Deaths_Suppressed
18,1999,Female,85-89,10.0,Not Applicable,Not Applicable,False
67,2000,Female,85-89,11.0,Not Applicable,Not Applicable,False
95,2000,Male,NS,10.0,Not Applicable,Not Applicable,False
116,2001,Female,85-89,10.0,Not Applicable,Not Applicable,False
165,2002,Female,85-89,10.0,Not Applicable,Not Applicable,False
214,2003,Female,85-89,20.0,Not Applicable,Not Applicable,False
312,2005,Female,85-89,23.0,Not Applicable,Not Applicable,False
313,2005,Female,90-94,11.0,Not Applicable,Not Applicable,False
336,2005,Male,85-89,11.0,Not Applicable,Not Applicable,False
361,2006,Female,85-89,24.0,Not Applicable,Not Applicable,False


In [79]:
not_applicable_deaths = df_clean[
    (df_clean['Population'] == 'Not Applicable') & 
    (df_clean['Deaths'] > 0.0)
]['Deaths'].sum()

total_deaths = df_clean['Deaths'].sum()

percentage = (not_applicable_deaths / total_deaths) * 100

print(f"Not Applicable deaths: {not_applicable_deaths:,.0f}")
print(f"Total deaths: {total_deaths:,.0f}")
print(f"Percentage: {percentage:.4f}%")

Not Applicable deaths: 959
Total deaths: 584,536
Percentage: 0.1641%


**959 'Not Applicable' deaths, or 0.1641% OF TOTAL DEATHS - SAFE TO DROP**

In [80]:
# return number of rows where 'Population' is 'Not Applicable
len(df_clean[df_clean['Population'] == 'Not Applicable'])

220

In [81]:
# keep only rows where 'Population' is not 'Not Applicable'
df_clean = df_clean[df_clean['Population'] != 'Not Applicable']

**ADD TO REUSABLE FUNCTION**

In [82]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 792 entries, 0 to 1070
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       792 non-null    int64  
 1   Sex                        792 non-null    str    
 2   Five-Year Age Groups Code  792 non-null    str    
 3   Deaths                     678 non-null    float64
 4   Population                 792 non-null    str    
 5   Crude Rate                 792 non-null    str    
 6   Deaths_Suppressed          792 non-null    bool   
dtypes: bool(1), float64(1), int64(1), str(4)
memory usage: 61.5 KB


In [83]:
# convert values in 'Population' column into numerical values
df_clean['Population'] = pd.to_numeric(df_clean['Population'], errors='coerce')
df_clean['Population'].dtype

dtype('int64')

In [84]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 792 entries, 0 to 1070
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       792 non-null    int64  
 1   Sex                        792 non-null    str    
 2   Five-Year Age Groups Code  792 non-null    str    
 3   Deaths                     678 non-null    float64
 4   Population                 792 non-null    int64  
 5   Crude Rate                 792 non-null    str    
 6   Deaths_Suppressed          792 non-null    bool   
dtypes: bool(1), float64(1), int64(2), str(3)
memory usage: 55.6 KB


### `Crude Rate`

In [85]:
# check unique values for 'Crude Rate'
df_clean['Crude Rate'].nunique()

222

In [86]:
# convert values in 'Crude Rate' column into numerical values
df_clean['Crude Rate'] = df_clean['Crude Rate'].astype('float64')

ValueError: could not convert string to float: 'Suppressed'

**UNEXPECTED ERROR - SOME VALUES CANNOT BE CONVERTED INTO NUMERICAL VALUE**

In [87]:
# return unique values in 'Crude Rate' column where NaN is False
(df_clean[pd.to_numeric(df_clean['Crude Rate'], errors='coerce')
    .isna()]['Crude Rate']
    .unique()
)

<ArrowStringArray>
['Suppressed', 'Unreliable']
Length: 2, dtype: str

**`Unreliable` AND `Suppressed` PREVENTS NUMERICAL CONVERSION**

In [88]:
# return sum counts of 'Suppressed' + 'Unreliable"
df_clean['Crude Rate'].isin(['Suppressed', 'Unreliable']).sum()

np.int64(187)

In [89]:
# return counts where 'Crude Rate' is either 'Unreliable' or 'Suppressed'
(df_clean[pd.to_numeric(df_clean['Crude Rate'], errors='coerce')
    .isna()]['Crude Rate']
    .value_counts()
)

Crude Rate
Suppressed    114
Unreliable     73
Name: count, dtype: int64

**TOTAL OF 187 INSTANCES WHERE 'Crude Rate' is 'Unreliable' or 'Suppressed**

In [90]:
# calculate percentage of 'Unreliable' + 'Suppressed' from 187 total rows
non_numeric = pd.to_numeric(df_clean['Crude Rate'], errors='coerce').isna().sum()

total = len(df_clean)

print(f"{non_numeric / total * 100:.2f}% of 'Crude Rate' is either 'Unreliable' or 'Suppressed'")

23.61% of 'Crude Rate' is either 'Unreliable' or 'Suppressed'


In [91]:
# create 'Crude Rate Calculated' column
df_clean['Crude Rate Calculated'] = (df_clean['Deaths'] / df_clean['Population']) * 100000

**ADD TO REUSUABLE FUNCTION**

In [92]:
# drop 'Crude Rate' column
df_clean.drop(columns = ['Crude Rate'], inplace = True)

**ADD TO REUSUABLE FUNCTION**

In [93]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 792 entries, 0 to 1070
Data columns (total 7 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       792 non-null    int64  
 1   Sex                        792 non-null    str    
 2   Five-Year Age Groups Code  792 non-null    str    
 3   Deaths                     678 non-null    float64
 4   Population                 792 non-null    int64  
 5   Deaths_Suppressed          792 non-null    bool   
 6   Crude Rate Calculated      678 non-null    float64
dtypes: bool(1), float64(2), int64(2), str(2)
memory usage: 51.7 KB


In [94]:
# reset index
df_clean = df_clean.reset_index(drop=True)

**ADD TO REUSUABLE FUNCTION**

## 7. Export cleaned data

In [95]:
print(f"Shape: {df_clean.shape}")
print(f"Year range: {df_clean['Year'].min()} – {df_clean['Year'].max()}")

# write to processed directory
df_clean.to_csv(
    '../data/processed/YEAR_GENDER_AGE_1999_2020_cleaned.csv', 
    index = False
)

print('Saved: ../data/processed/YEAR_GENDER_AGE_1999_2020_cleaned.csv')

Shape: (792, 7)
Year range: 1999 – 2020
Saved: ../data/processed/YEAR_GENDER_AGE_1999_2020_cleaned.csv


In [96]:
os.path.exists('../data/processed/YEAR_GENDER_AGE_1999_2020_cleaned.csv')

True